lib/endpoints/class-wp-json-attachments-controller.php

```diff
@@ -0,0 +1,337 @@
+<?php
+
+class WP_JSON_Attachments_Controller extends WP_JSON_Posts_Controller {
+
+	/**
+	 * Create a single attachment
+	 *
+	 * @param WP_JSON_Request $request Full details about the request
+	 * @return WP_Error|WP_HTTP_ResponseInterface
+	 */
+	public function create_item( $request ) {
+
+		// Permissions check - Note: "upload_files" cap is returned for an attachment by $post_type_obj->cap->create_posts
+		$post_type_obj = get_post_type_object( $this->post_type );
+		if ( ! current_user_can( $post_type_obj->cap->create_posts ) || ! current_user_can( $post_type_obj->cap->edit_posts ) ) {
+			return new WP_Error( 'json_cannot_create', __( 'Sorry, you are not allowed to post on this site.' ), array( 'status' => 400 ) );
+		}
+
+		// If a user is trying to attach to a post make sure they have permissions. Bail early if post_id is not being passed
+		if ( ! empty( $request['post_id'] ) ) {
+			$parent = get_post( (int) $request['post_id'] );
+			$post_parent_type = get_post_type_object( $parent->post_type );
+			if ( ! current_user_can( $post_parent_type->cap->edit_post, $request['post_id'] ) ) {
+				return new WP_Error( 'json_cannot_edit', __( 'Sorry, you are not allowed to edit this post.' ), array( 'status' => 401 ) );
+			}
+		}
+
+		// Get the file via $_FILES or raw data
+		$files = $request->get_file_params();
+		$headers = $request->get_headers();
+		if ( ! empty( $files ) ) {
+			$file = $this->upload_from_file( $files, $headers );
+		} else {
+			$file = $this->upload_from_data( $request->get_body(), $headers );
+		}
+
+		if ( is_wp_error( $file ) ) {
+			return $file;
+		}
+
+		$name       = basename( $file['file'] );
+		$name_parts = pathinfo( $name );
+		$name       = trim( substr( $name, 0, -(1 + strlen( $name_parts['extension'] ) ) ) );
+
+		$url     = $file['url'];
+		$type    = $file['type'];
+		$file    = $file['file'];
+		$title   = $name;
+		$caption = '';
+
+		// use image exif/iptc data for title and caption defaults if possible
+		if ( $image_meta = @wp_read_image_metadata( $file ) ) {
+			if ( empty( $request['title'] ) && trim( $image_meta['title'] ) && ! is_numeric( sanitize_title( $image_meta['title'] ) ) ) {
+				$title = $image_meta['title'];
+			}
+
+			if ( empty( $request['caption'] ) && trim( $image_meta['caption'] ) ) {
+				$caption = $image_meta['caption'];
+			}
+		}
+
+		$attachment = $this->prepare_item_for_database( $request );
+		$attachment->file = $file;
+		$attachment->post_mime_type = $type;
+		$attachment->guid = $url;
+		$id = wp_insert_post( $attachment, true );
```

Comment:
I used `wp_insert_post()` because there's no functional reason to use `wp_insert_attachment()` and `wp_insert_post()` supports returning `WP_Error` when there's a failure (and `wp_insert_attachment()` does not).